File: CombinationSum-Backtrack.java
class Solution {
    List<List<Integer>> output;
    int sum;
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        
        output = new ArrayList<>();
        List<Integer> currentPath = new ArrayList<>();
        sum = target;
        if (candidates == null || candidates.length == 0) return output;
        helper(candidates,0,currentPath,0);
        return output;
        
    }
    
    private void helper(int[] candidates, int currSum, List<Integer> currentPath,int index ){
        
        //Base Case
        
       if (currSum == sum){
            output.add(new ArrayList<>(currentPath));
           return;
        }
        
        if (currSum > sum || index == candidates.length) return;
        
        for (int x = index; x < candidates.length; x++){
            
            // action
            currentPath.add(candidates[x]);
            
            //recursion
            
            helper(candidates, currSum + candidates[x], currentPath, x);
            
            //backtrack
            
            currentPath.remove(currentPath.size() - 1);
            
        }

        //dont chose
        //helper(candidates,currSum, new ArrayList<>(currentPath),index+1);
        
        //chose
        //currentPath.add(candidates[index]);
        //helper(candidates,currSum + candidates[index], new ArrayList<>(currentPath),index);
        
       
    }
}

File: combination_sum_backtrack.py
class Solution:
	def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
		result = []
		if not candidates:
			return result
		self.backtrack(candidates, target, [], 0, 0, result)
		return result
	
	def backtrack(self, candidates, target, temp, index, curr_sum, result):
		if curr_sum>target:
			return
		if curr_sum == target:
			result.append(temp.copy())
			return
		
		for i in range(index, len(candidates)):
			#action
			temp.append(candidates[i])
			#recurse
			self.backtrack(candidates, target, temp, i, curr_sum+candidates[i], result)
			#backtrack
			temp.pop()

#time complexity - O(2^n) as at every step we have 2 choices to choose or not choose

#space complexity - O(2^n) as we maintain a temp at every stage

#all test cases passed